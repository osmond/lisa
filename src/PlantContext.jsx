import { createContext, useContext, useEffect, useState } from "react";
import initialPlants from "./plants.json";
import { useWeather } from "./WeatherContext.jsx";
import { getNextWateringDate } from "./utils/watering.js";
import { getWaterPlan, getSmartWaterPlan } from "./utils/waterCalculator.js";
import { cubicInchesToMl, mlToOz } from "./utils/volume.js";
import autoTag from "./utils/autoTag.js";

const PlantContext = createContext();

export const addBase = (url) => {
  if (!url) return url;
  if (/^https?:/.test(url) || url.startsWith("data:")) return url;
  const base = (process.env.VITE_BASE_PATH || "/").replace(/\/$/, "");
  if (url.startsWith(base)) return url;
  return `${base}${url.startsWith("/") ? "" : "/"}${url}`;
};

const mapPhoto = (photo) => {
  if (!photo) return photo;
  if (typeof photo === "string") {
    return { src: addBase(photo), caption: "", tags: [] };
  }
  return { ...photo, src: addBase(photo.src), tags: photo.tags || [] };
};

export function PlantProvider({ children }) {
  const [plants, setPlants] = useState(() => {
    const mapPlant = (p) => {
      const wp = p.waterPlan || { interval: 0 };
      const ml =
        wp.volume_ml !== undefined
          ? wp.volume_ml
          : cubicInchesToMl(wp.volume || 0);
      const oz =
        wp.volume_oz !== undefined ? wp.volume_oz : mlToOz(ml);
      const smart = p.smartWaterPlan
        ? {
            ...p.smartWaterPlan,
            volume_ml:
              p.smartWaterPlan.volume_ml !== undefined
                ? p.smartWaterPlan.volume_ml
                : cubicInchesToMl(p.smartWaterPlan.volume || 0),
            volume_oz:
              p.smartWaterPlan.volume_oz !== undefined
                ? p.smartWaterPlan.volume_oz
                : mlToOz(
                    p.smartWaterPlan.volume_ml !== undefined
                      ? p.smartWaterPlan.volume_ml
                      : cubicInchesToMl(p.smartWaterPlan.volume || 0),
                  ),
          }
        : null;
      return {
        ...p,
        image: addBase(p.image),
        photos: (p.photos || p.gallery || []).map(mapPhoto),
        careLog: (p.careLog || []).map((ev) => ({ ...ev, tags: ev.tags || [] })),
        diameter: p.diameter || 0,
        petSafe: !!p.petSafe,
        waterPlan: { interval: wp.interval || 0, volume_ml: Math.round(ml), volume_oz: Math.round(oz) },
        carePlan: p.carePlan || null,
        smartWaterPlan: smart,
        lastEvaluated: p.lastEvaluated || null,
      };
    };

    if (typeof localStorage !== "undefined") {
      const stored = localStorage.getItem("plants");
      if (stored) {
        try {
          return JSON.parse(stored).map(mapPlant);
        } catch {
          // fall through to initial plants
        }
      }
    }
    return initialPlants.map(mapPlant);
  });

  useEffect(() => {
    let ignore = false;
    async function load() {
      if (typeof fetch === 'undefined') return;
      try {
        const res = await fetch("/api/plants");
        if (!res.ok) return;
        const data = await res.json();
        if (!ignore) setPlants(data.map(mapPlant));
      } catch {
        // ignore network errors and keep local data
      }
    }
    load();
    return () => {
      ignore = true;
    };
  }, []);

  const weatherCtx = useWeather();
  const weather = { rainTomorrow: weatherCtx?.forecast?.rainfall || 0 };

  useEffect(() => {
    if (typeof localStorage !== "undefined") {
      localStorage.setItem("plants", JSON.stringify(plants));
    }
  }, [plants]);

  const evaluatePlans = (forecast) => {
    if (!forecast) return;
    setPlants((prev) =>
      prev.map((p) => {
        if (!p.name || !p.diameter) return p;
        const logs = (p.careLog || []).filter((l) => l.type === "Watered");
        const raw = getSmartWaterPlan(
          { name: p.name, diameter: p.diameter, light: p.light },
          forecast,
          logs,
        );
        const ml = cubicInchesToMl(raw.volume);
        const plan = {
          interval: raw.interval,
          reason: raw.reason,
          volume_ml: Math.round(ml),
          volume_oz: Math.round(mlToOz(ml)),
        };
        return {
          ...p,
          smartWaterPlan: plan,
          lastEvaluated: new Date().toISOString(),
        };
      }),
    );
  };

  useEffect(() => {
    const handler = (e) => evaluatePlans(e.detail);
    window.addEventListener("weather-updated", handler);
    return () => window.removeEventListener("weather-updated", handler);
  }, []);

  const [timelineNotes, setTimelineNotes] = useState(() => {
    if (typeof localStorage !== "undefined") {
      const stored = localStorage.getItem("timelineNotes");
      if (stored) {
        try {
          return JSON.parse(stored).map((n) => ({ ...n, tags: n.tags || [] }));
        } catch {
          // ignore
        }
      }
    }
    return [];
  });

  useEffect(() => {
    if (typeof localStorage !== "undefined") {
      localStorage.setItem("timelineNotes", JSON.stringify(timelineNotes));
    }
  }, [timelineNotes]);

  const addTimelineNote = async (text) => {
    const date = new Date().toISOString().slice(0, 10);
    const tags = await autoTag(text);
    setTimelineNotes((prev) => [...prev, { date, text, tags }]);
  };

  const logEvent = async (id, type, note = "") => {
    const date = new Date().toISOString().slice(0, 10);
    const tags = await autoTag(note);
    setPlants((prev) =>
      prev.map((p) =>
        p.id === id
          ? {
              ...p,
              careLog: [...(p.careLog || []), { date, type, note, tags }],
            }
          : p,
      ),
    );
  };

  const markWatered = async (id, note) => {
    const tags = await autoTag(note);
    const today = new Date().toISOString().slice(0, 10);
    setPlants((prev) =>
      prev.map((p) => {
        if (p.id === id) {
          const { date: nextStr, reason } = getNextWateringDate(today, weather);
          const newLog = [
            ...(p.careLog || []),
            { date: today, type: "Watered", note, tags },
          ];
          const raw = getSmartWaterPlan(
            { name: p.name, diameter: p.diameter, light: p.light },
            weatherCtx?.forecast,
            newLog.filter((l) => l.type === "Watered"),
          );
          const volMl = cubicInchesToMl(raw.volume);
          const plan = {
            interval: raw.interval,
            reason: raw.reason,
            volume_ml: Math.round(volMl),
            volume_oz: Math.round(mlToOz(volMl)),
          };
          return {
            ...p,
            careLog: newLog,
            lastWatered: today,
            nextWater: nextStr,
            nextWaterReason: reason,
            smartWaterPlan: plan,
            lastEvaluated: new Date().toISOString(),
          };
        }
        return p;
      }),
    );
  };

  const markFertilized = (id, note) => {
    setPlants((prev) =>
      prev.map((p) => {
        if (p.id === id) {
          const today = new Date().toISOString().slice(0, 10);
          let nextDate = new Date(today);
          if (p.lastFertilized && p.nextFertilize) {
            const last = new Date(p.lastFertilized);
            const next = new Date(p.nextFertilize);
            const diff = Math.round((next - last) / 86400000) || 30;
            nextDate.setDate(nextDate.getDate() + diff);
          } else {
            nextDate.setMonth(nextDate.getMonth() + 1);
          }
          return {
            ...p,
            lastFertilized: today,
            nextFertilize: nextDate.toISOString().slice(0, 10),
          };
        }
        return p;
      }),
    );
    logEvent(id, "Fertilized", note);
  };

  const addPlant = (plant) => {
    setPlants((prev) => {
      const nextId = prev.reduce((m, p) => Math.max(m, p.id), 0) + 1;

      const newPlant = {
        id: nextId,
        photos: [],
        careLog: [],
        diameter: 0,
        waterPlan: { interval: 0, volume_ml: 0, volume_oz: 0 },
        carePlan: null,
        ...plant,
      };
      if (typeof fetch !== 'undefined')
        fetch('/api/plants', {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify(newPlant),
        }).catch(() => {})
      return [...prev, newPlant];
    });
  };

  const updatePlant = (id, updates) => {
    setPlants((prev) =>
      prev.map((p) => {
        if (p.id !== id) return p;
        const next = { ...p, ...updates };
        if (Object.prototype.hasOwnProperty.call(updates, "diameter")) {
          const base = getWaterPlan(next.name, updates.diameter, next.light);
          const ml = cubicInchesToMl(base.volume);
          next.waterPlan = {
            interval: base.interval,
            volume_ml: Math.round(ml),
            volume_oz: Math.round(mlToOz(ml)),
          };
        }
        return next;
      }),
    );
    if (typeof fetch !== 'undefined')
      fetch(`/api/plants/${id}`, {
        method: 'PUT',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(updates),
      }).catch(() => {})
  };

  const removePlant = (id) => {
    setPlants((prev) => prev.filter((p) => p.id !== id));
    if (typeof fetch !== 'undefined')
      fetch(`/api/plants/${id}`, { method: 'DELETE' }).catch(() => {})
  };

  const restorePlant = (plant, index) => {
    setPlants((prev) => {
      const arr = [...prev];
      const pos = index != null ? index : arr.length;
      arr.splice(pos, 0, plant);
      return arr;
    });
  };

  const addPhoto = async (id, photo) => {
    const newPhoto = mapPhoto(photo);
    const tags = await autoTag(newPhoto.caption || "");
    newPhoto.tags = tags;
    setPlants((prev) =>
      prev.map((p) =>
        p.id === id ? { ...p, photos: [...(p.photos || []), newPhoto] } : p,
      ),
    );
    const form = new FormData();
    if (photo.file) form.append('photos', photo.file);
    if (typeof fetch !== 'undefined' && form.has('photos'))
      fetch(`/api/plants/${id}/photos`, { method: 'POST', body: form }).catch(
        () => {},
      );
  };

  const removePhoto = (id, index) => {
    setPlants((prev) =>
      prev.map((p) => {
        if (p.id === id) {
          const updated = (p.photos || []).filter((_, i) => i !== index);
          return { ...p, photos: updated };
        }
        return p;
      }),
    );
    if (typeof fetch !== 'undefined')
      fetch(`/api/plants/${id}/photos/${index}`, { method: 'DELETE' }).catch(() => {});
  };

  return (
    <PlantContext.Provider
      value={{
        plants,
        markWatered,
        markFertilized,
        logEvent,
        addPlant,
        updatePlant,
        removePlant,
        restorePlant,
        addPhoto,
        removePhoto,
        timelineNotes,
        addTimelineNote,
      }}
    >
      {children}
    </PlantContext.Provider>
  );
}

export const usePlants = () => useContext(PlantContext);
