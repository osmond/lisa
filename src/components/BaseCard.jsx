import React from 'react'

export default function BaseCard({ variant = 'task', className = '', children, ...props }) {
  const base = 'rounded-2xl'
  const variants = {
    task: '',
    summary: 'p-4 border dark:border-gray-600 bg-white dark:bg-gray-700',
  }
  const variantClass = variants[variant] ?? ''
  return (
    <div className={`${base} ${variantClass} ${className}`} {...props}>
      {children}
    </div>
  )
}
