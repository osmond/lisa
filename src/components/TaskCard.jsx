import { Drop, Sun, PencilSimpleLine, ClockCounterClockwise, Trash } from 'phosphor-react'
import { useNavigate } from 'react-router-dom'
import { getWateringInfo } from '../utils/watering.js'
import { usePlants } from '../PlantContext.jsx'
import useSnackbar from '../hooks/useSnackbar.jsx'
import Badge from './Badge.jsx'
import useSwipe from '../hooks/useSwipe.js'

export default function TaskCard({
  task,
  urgent = false,
  completed = false,
  compact = false,
  swipeable = true,
}) {
  const { daysSince, eto } = getWateringInfo(task.lastWatered, { eto: task.eto })
  const navigate = useNavigate()
  const { plants, markWatered, markFertilized, updatePlant } = usePlants()
  const { Snackbar, showSnackbar } = useSnackbar()

  const goToDetail = e => {
    e.stopPropagation()
    const room = task.room ? encodeURIComponent(task.room) : null
    navigate(room ? `/room/${room}/plant/${task.plantId}` : `/plant/${task.plantId}`)
  }

  const handleComplete = () => {
    const prev = plants.find(p => p.id === task.plantId)
    if (task.type === 'Water') {
      markWatered(task.plantId, '')
      showSnackbar('Watered', () => updatePlant(task.plantId, prev))
    } else if (task.type === 'Fertilize') {
      markFertilized(task.plantId, '')
      showSnackbar('Fertilized', () => updatePlant(task.plantId, prev))
    }
  }

  const handleEdit = () => {
    const room = task.room ? encodeURIComponent(task.room) : null
    navigate(
      room ? `/room/${room}/plant/${task.plantId}/edit` : `/plant/${task.plantId}/edit`
    )
  }

  const handleReschedule = () => {
    const prev = plants.find(p => p.id === task.plantId)
    const next = new Date(task.date || new Date())
    next.setDate(next.getDate() + 1)
    const updates =
      task.type === 'Water'
        ? { nextWater: next.toISOString().slice(0, 10) }
        : { nextFertilize: next.toISOString().slice(0, 10) }
    updatePlant(task.plantId, updates)
    showSnackbar('Rescheduled', () => updatePlant(task.plantId, prev))
  }

  const handleDelete = () => {
    const prev = plants.find(p => p.id === task.plantId)
    const updates =
      task.type === 'Water'
        ? { nextWater: null }
        : { nextFertilize: null }
    updatePlant(task.plantId, updates)
    showSnackbar('Deleted', () => updatePlant(task.plantId, prev))
  }

  const { dx, start, move, end } = useSwipe(
    diff => {
      if (!swipeable) return
      if (diff > 60) {
        handleComplete()
        navigator.vibrate?.(10)
      } else if (diff < -60) {
        const room = task.room ? encodeURIComponent(task.room) : null
        navigate(
          room ? `/room/${room}/plant/${task.plantId}` : `/plant/${task.plantId}`
        )
      }
    },
    { threshold: 30 }
  )

  const showActionBar = dx < 0 && dx > -60

  return (
    <>
    <div
      data-testid="task-card"
      tabIndex="0"
      aria-label={`Task card for ${task.plantName}`}
      className={`relative flex items-center p-5 gap-4 rounded-2xl shadow border border-neutral-200 dark:border-gray-600 ${completed ? 'bg-gray-100 dark:bg-gray-800 opacity-50' : 'bg-neutral-50 dark:bg-gray-700'}${urgent ? ' ring-2 ring-green-300 dark:ring-green-400' : ''}`}
      style={{ transform: `translateX(${swipeable ? dx : 0}px)`, transition: dx === 0 ? 'transform 0.2s' : 'none' }}
      onPointerDown={start}
      onPointerMove={move}
      onPointerUp={end}
      onPointerCancel={end}
    >
      <div
        className={`task-action-bar ${showActionBar ? 'show' : ''}`}
        role="group"
        aria-label="Task actions"
      >
        <button
          type="button"
          aria-label="Edit task"
          onClick={handleEdit}
          className="task-action bg-blue-600 text-white"
        >
          <PencilSimpleLine className="w-4 h-4" aria-hidden="true" />
          Edit
        </button>
        <button
          type="button"
          aria-label="Reschedule task"
          onClick={handleReschedule}
          className="task-action bg-yellow-600 text-white"
        >
          <ClockCounterClockwise className="w-4 h-4" aria-hidden="true" />
          Reschedule
        </button>
        <button
          type="button"
          aria-label="Delete task"
          onClick={handleDelete}
          className="task-action bg-red-600 text-white"
        >
          <Trash className="w-4 h-4" aria-hidden="true" />
          Delete
        </button>
      </div>
      <div className="flex items-center flex-1 gap-4">
      <button
        type="button"
        onClick={goToDetail}
        className={`w-16 h-16 rounded-full flex items-center justify-center shadow-sm bg-neutral-100 dark:bg-gray-800 ${
            task.type === 'Water'
              ? 'ring-2 ring-water-300'
              : task.type === 'Fertilize'
              ? 'ring-2 ring-fertilize-300'
              : 'ring-2 ring-healthy-300'
          }`}
      >
        <img
          src={task.image}
          alt={task.plantName}
          className="w-12 h-12 rounded-full object-cover"
        />
      </button>
        <div className="w-px self-stretch bg-gray-200 dark:bg-gray-600" aria-hidden="true" />
        <div className="flex-1 min-w-0">
          <div className="text-base font-semibold text-gray-900 dark:text-gray-100 truncate">
            {task.plantName}
          </div>
          <div className="flex items-center gap-2 mt-1">
            <Badge
              Icon={
                task.type === 'Water'
                  ? Drop
                  : task.type === 'Fertilize'
                  ? Sun
                  : undefined
              }
              variant={completed ? 'complete' : 'info'}
            >
              {completed
                ? task.type === 'Water'
                  ? 'Watered!'
                  : task.type === 'Fertilize'
                  ? 'Fertilized!'
                  : task.type
                : task.type === 'Water'
                ? 'To Water'
                : task.type === 'Fertilize'
                ? 'To Fertilize'
                : task.type}
            </Badge>
            {daysSince != null && (
              <span className="text-timestamp text-gray-400">
                {daysSince} {daysSince === 1 ? 'day' : 'days'} since care
              </span>
            )}
          </div>
          {!compact && task.reason && (
            <p className="text-xs text-gray-500 font-body mt-0.5">{task.reason}</p>
          )}
        </div>
      </div>
          {completed && (
            <div className="absolute inset-0 flex items-center justify-center pointer-events-none task-complete-fade">
              <svg
                className="w-8 h-8 text-healthy-600 check-pop swipe-check"
                xmlns="http://www.w3.org/2000/svg"
                fill="none"
                viewBox="0 0 24 24"
                strokeWidth="1.5"
                stroke="currentColor"
                aria-hidden="true"
              >
                <path strokeLinecap="round" strokeLinejoin="round" d="M4.5 12.75l6 6 9-13.5" />
              </svg>
            </div>
          )}
          {!compact && (
            <div className="mt-2">
              <span
                className="px-2 py-0.5 text-sm rounded-xl bg-gray-200 dark:bg-gray-600 text-gray-800 dark:text-gray-100 opacity-70"
                aria-label={`Evapotranspiration (ET₀): ${eto ?? 'N/A'} mm | Last watered ${daysSince ?? '?'} days ago`}
                title="Evapotranspiration (ET₀) is water lost from soil and plants"
              >
                Evapotranspiration (ET₀): {eto ?? '—'} mm | Last watered {daysSince ?? '?'} days ago
              </span>
            </div>
          )}
        </div>
      <Snackbar />
    </>
  )
}

