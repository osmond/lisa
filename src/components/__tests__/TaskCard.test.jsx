import { render, screen, fireEvent, act } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { MemoryRouter, Routes, Route } from 'react-router-dom'
import TaskCard from '../TaskCard.jsx'
import BaseCard from '../BaseCard.jsx'

import { PlantProvider, usePlants } from '../../PlantContext.jsx'

beforeAll(() => {
  if (typeof PointerEvent === 'undefined') {
    window.PointerEvent = window.MouseEvent
  }
})

jest.mock('../../PlantContext.jsx', () => {
  const actual = jest.requireActual('../../PlantContext.jsx')
  return { ...actual, usePlants: jest.fn() }
})


const usePlantsMock = usePlants

const markWatered = jest.fn()
const markFertilized = jest.fn()

beforeEach(() => {
  markWatered.mockClear()
  markFertilized.mockClear()
  usePlantsMock.mockReturnValue({
    plants: [],
    markWatered,
    markFertilized,
    logEvent: jest.fn(),
    addPlant: jest.fn(),
    updatePlant: jest.fn(),
    removePlant: jest.fn(),
    addPhoto: jest.fn(),
    removePhoto: jest.fn(),
  })
})

const task = {
  id: 1,
  plantId: 1,
  plantName: 'Monstera',
  image: 'https://images.pexels.com/photos/5699660/pexels-photo-5699660.jpeg',
  type: 'Water',
  lastWatered: '2025-07-07'
}

test('renders task text', () => {
  render(
    <MemoryRouter>
      <BaseCard variant="task">
        <TaskCard task={task} />
      </BaseCard>
    </MemoryRouter>
  )
  expect(screen.getByText('Monstera')).toBeInTheDocument()
  const badge = screen.getByText('To Water')
  expect(badge).toBeInTheDocument()
  expect(badge).toHaveClass('inline-flex')
})

test('incomplete tasks show alert style', () => {
  const { container } = render(
    <PlantProvider>
      <MemoryRouter>
        <BaseCard variant="task">
          <TaskCard task={task} />
        </BaseCard>
      </MemoryRouter>
    </PlantProvider>
  )
  const wrapper = container.querySelector('[data-testid="task-card"]')
  expect(wrapper).toHaveClass('bg-sage')
  expect(wrapper).toHaveClass('ring-accent')
})

test('applies highlight when urgent', () => {
  const { container } = render(
    <MemoryRouter>
      <BaseCard variant="task">
        <TaskCard task={task} urgent />
      </BaseCard>
    </MemoryRouter>
  )
  const wrapper = container.querySelector('[data-testid="task-card"]')
  expect(wrapper).toHaveClass('ring-2')
  expect(wrapper).toHaveClass('ring-green-300')
  expect(wrapper).toHaveClass('dark:ring-green-400')
})

test('applies overdue styling', () => {
  const { container } = render(
    <MemoryRouter>
      <BaseCard variant="task">
        <TaskCard task={task} overdue />
      </BaseCard>
    </MemoryRouter>
  )
  const wrapper = container.querySelector('[data-testid="task-card"]')
  expect(wrapper).toHaveClass('ring-2')
  expect(wrapper).toHaveClass('ring-orange-300')
  expect(screen.getByTestId('overdue-badge')).toBeInTheDocument()
})

test('shows completed state', () => {
  const { container } = render(
    <MemoryRouter>
      <BaseCard variant="task">
        <TaskCard task={task} completed />
      </BaseCard>
    </MemoryRouter>
  )
  const wrapper = container.querySelector('[data-testid="task-card"]')
  expect(wrapper).toHaveClass('opacity-50')
  expect(wrapper).toHaveClass('bg-gray-100')
  const checkbox = container.querySelector('input[type="checkbox"]')
  expect(checkbox).toBeChecked()
  expect(container.querySelector('.check-pop')).toBeInTheDocument()
})

test('icon svg is aria-hidden', () => {
  const { container } = render(
    <MemoryRouter>
      <BaseCard variant="task">
        <TaskCard task={task} />
      </BaseCard>
    </MemoryRouter>
  )
  const svg = container.querySelector('svg')
  expect(svg).toHaveAttribute('aria-hidden', 'true')
})

test('shows info chip with accessibility label', () => {
  jest.useFakeTimers().setSystemTime(new Date('2025-07-10'))
  render(
    <MemoryRouter>
      <BaseCard variant="task">
        <TaskCard task={task} />
      </BaseCard>
    </MemoryRouter>
  )
  const chip = screen.getByText(/ETâ‚€/i)
  expect(chip).toHaveAttribute(
    'aria-label',
    expect.stringContaining('Last watered 3 days ago')
  )
  jest.useRealTimers()
})

test('compact mode hides reason and ETâ‚€ info', () => {
  const compactTask = { ...task, reason: 'Needs water' }
  render(
    <MemoryRouter>
      <BaseCard variant="task">
        <TaskCard task={compactTask} compact />
      </BaseCard>
    </MemoryRouter>
  )
  expect(screen.queryByText('Needs water')).not.toBeInTheDocument()
  expect(screen.queryByText(/ETâ‚€/)).not.toBeInTheDocument()
})

test('mark as done does not navigate', () => {
  render(
    <MemoryRouter initialEntries={['/']}>
      <Routes>
        <Route
          path="/"
          element={(
            <BaseCard variant="task">
              <TaskCard task={task} />
            </BaseCard>
          )}
        />
        <Route path="/plant/:id" element={<div>Plant Page</div>} />
      </Routes>
    </MemoryRouter>
  )
  fireEvent.click(screen.getByRole('checkbox'))
  const dialog = screen.getByRole('dialog', { name: /optional note/i })
  fireEvent.change(dialog.querySelector('textarea'), { target: { value: '' } })
  fireEvent.click(screen.getByText('Save'))
  expect(screen.queryByText('Plant Page')).not.toBeInTheDocument()
})

test('completing with note calls log function', () => {
  render(
    <MemoryRouter>
      <BaseCard variant="task">
        <TaskCard task={task} />
      </BaseCard>
    </MemoryRouter>
  )
  fireEvent.click(screen.getByRole('checkbox'))
  const dialog = screen.getByRole('dialog', { name: /optional note/i })
  fireEvent.change(dialog.querySelector('textarea'), { target: { value: 'hello' } })
  fireEvent.click(screen.getByText('Save'))
  expect(markWatered).toHaveBeenCalledWith(1, 'hello')
})

test('clicking card adds ripple effect', () => {
  const { container } = render(
    <MemoryRouter>
      <BaseCard variant="task">
        <TaskCard task={task} />
      </BaseCard>
    </MemoryRouter>
  )
  const wrapper = container.querySelector('[data-testid="task-card"]')
  fireEvent.mouseDown(wrapper)
  expect(container.querySelector('.ripple-effect')).toBeInTheDocument()
})

test('keyboard Enter and Space trigger completion with ripple', async () => {
  const onComplete = jest.fn()
  const { container } = render(
    <MemoryRouter>
      <BaseCard variant="task">
        <TaskCard task={task} onComplete={onComplete} />
      </BaseCard>
    </MemoryRouter>
  )
  const button = screen.getByRole('button', { name: /mark complete/i })
  const user = userEvent.setup()
  button.focus()
  await user.keyboard('{Enter}')
  expect(container.querySelector('.ripple-effect')).toBeInTheDocument()
  expect(onComplete).toHaveBeenCalledWith(task)
  container.querySelector('.ripple-effect')?.remove()
  await user.keyboard(' ')
  expect(container.querySelector('.ripple-effect')).toBeInTheDocument()
  expect(onComplete).toHaveBeenCalledTimes(2)
})

test('arrow right marks task complete', () => {
  const onComplete = jest.fn()
  render(
    <MemoryRouter>
      <BaseCard variant="task">
        <TaskCard task={task} onComplete={onComplete} />
      </BaseCard>
    </MemoryRouter>
  )
  const wrapper = screen.getByTestId('task-card')
  wrapper.focus()
  fireEvent.keyDown(wrapper, { key: 'ArrowRight' })
  expect(onComplete).toHaveBeenCalledWith(task)
})

test('swipe right marks task complete', async () => {
  const onComplete = jest.fn()
  render(
    <MemoryRouter>
      <BaseCard variant="task">
        <TaskCard task={task} onComplete={onComplete} />
      </BaseCard>
    </MemoryRouter>
  )
  const wrapper = screen.getByTestId('task-card')
  fireEvent.pointerDown(wrapper, { clientX: 0, buttons: 1 })
  fireEvent.pointerMove(wrapper, { clientX: 80, buttons: 1 })
  fireEvent.pointerUp(wrapper, { clientX: 80 })
  const user = userEvent.setup()
  await act(async () => {
    fireEvent.touchStart(wrapper, { touches: [{ clientX: 0 }] })
    fireEvent.touchMove(wrapper, { touches: [{ clientX: 80 }] })
    fireEvent.touchEnd(wrapper)
  })
  expect(onComplete).toHaveBeenCalledWith(task)
})

test('matches snapshot in dark mode', () => {
  jest.useFakeTimers().setSystemTime(new Date('2025-07-16'))
  document.documentElement.classList.add('dark')
  const { container } = render(
    <MemoryRouter>
      <BaseCard variant="task">
        <TaskCard task={task} urgent />
      </BaseCard>
    </MemoryRouter>
  )
  expect(container.firstChild).toMatchSnapshot()
  document.documentElement.classList.remove('dark')
  jest.useRealTimers()
})

test('shows toast on completion', () => {
  jest.useFakeTimers()
  render(
    <MemoryRouter>
      <BaseCard variant="task">
        <TaskCard task={task} />
      </BaseCard>
    </MemoryRouter>
  )
  fireEvent.click(screen.getByRole('checkbox'))
  const dialog = screen.getByRole('dialog', { name: /optional note/i })
  fireEvent.change(dialog.querySelector('textarea'), { target: { value: '' } })
  fireEvent.click(screen.getByText('Save'))
  expect(screen.getByText('Watered Monstera ğŸŒ¿')).toBeInTheDocument()
  act(() => {
    jest.runAllTimers()
  })
  expect(screen.queryByText('Watered Monstera ğŸŒ¿')).not.toBeInTheDocument()
  jest.useRealTimers()
})
