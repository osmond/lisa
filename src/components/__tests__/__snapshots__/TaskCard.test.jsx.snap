// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot in dark mode 1`] = `
<div
  class="  "
>
  <div
    aria-label="Task card for Monstera"
    class="relative overflow-hidden min-h-[130px] rounded-xl"
    data-testid="task-card"
    tabindex="0"
  >
    <div
      class="relative flex items-center gap-4 shadow-md bg-white dark:bg-gray-700 ring-2 ring-green-300 dark:ring-green-400"
      style="transform: translateX(0px); transition: transform 0.2s;"
    >
      <div
        class="flex items-center flex-1 gap-4"
      >
        <div
          class="w-16 h-16 rounded-full flex items-center justify-center bg-green-50 dark:bg-gray-800 ring-2 ring-water-300"
        >
          <img
            alt="Monstera"
            class="w-12 h-12 rounded-full object-cover"
            src="https://images.pexels.com/photos/5699660/pexels-photo-5699660.jpeg"
          />
        </div>
        <div
          aria-hidden="true"
          class="w-px self-stretch bg-gray-200 dark:bg-gray-600"
        />
        <div
          class="flex-1 min-w-0"
        >
          <div
            class="flex items-center justify-between gap-2"
          >
            <p
              class="font-semibold text-lg text-gray-900 dark:text-gray-100 truncate"
            >
              Monstera
            </p>
          </div>
          <div
            class="text-sm flex flex-wrap items-center gap-1 text-gray-400 mt-0.5"
          >
            <span
              class="inline-flex items-center gap-1 px-2 py-0.5 rounded-full shadow-sm text-sm font-medium bg-water-100/90 text-water-800"
            >
              <svg
                aria-hidden="true"
                class="w-3 h-3"
                fill="currentColor"
                height="1em"
                viewBox="0 0 256 256"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  fill="none"
                  height="256"
                  width="256"
                />
                <path
                  d="M208,144c0-72-80-128-80-128S48,72,48,144a80,80,0,0,0,160,0Z"
                  fill="none"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="16"
                />
                <path
                  d="M136.1,191.2a47.9,47.9,0,0,0,39.2-39.1"
                  fill="none"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="16"
                />
              </svg>
              To Water
            </span>
            <span
              class="text-xs text-gray-400"
            >
              9
               
              days
               since care
            </span>
          </div>
        </div>
      </div>
      <div
        class="mt-2"
      >
        <span
          aria-label="Evapotranspiration (ET₀): N/A mm | Last watered 9 days ago"
          class="px-2 py-0.5 text-sm rounded-xl bg-gray-200 dark:bg-gray-600 text-gray-800 dark:text-gray-100 opacity-70"
          title="Evapotranspiration (ET₀) is water lost from soil and plants"
        >
          Evapotranspiration (ET₀): 
          —
           mm | Last watered 
          9
           days ago
        </span>
      </div>
    </div>
  </div>
</div>
`;
